cmd/kubeapps-apis/server/plugins.go

```diff
@@ -232,3 +253,150 @@ func listSOFiles(fsys fs.FS, pluginDirs []string) ([]string, error) {
+			if configs.KubeappsClusterName == "" {
+				configs.KubeappsClusterName = c.Name
+			} else {
+				return kube.ClustersConfig{}, nil, fmt.Errorf("only one cluster can be configured using either 'isKubeappsCluster: true' or without an apiServiceURL to refer to the cluster on which Kubeapps is installed, two defined: %q, %q", configs.KubeappsClusterName, c.Name)
+			}
+		}
+
+		// We need to decode the base64-encoded cadata from the input.
+		if c.CertificateAuthorityData != "" {
+			decodedCAData, err := base64.StdEncoding.DecodeString(c.CertificateAuthorityData)
+			if err != nil {
+				return kube.ClustersConfig{}, deferFn, err
+			}
+			c.CertificateAuthorityDataDecoded = string(decodedCAData)
+
+			// We also need a CAFile field because Helm uses the genericclioptions.ConfigFlags
+			// struct which does not support CAData.
+			// https://github.com/kubernetes/cli-runtime/issues/8
+			c.CAFile = filepath.Join(caFilesDir, c.Name)
+			err = ioutil.WriteFile(c.CAFile, decodedCAData, 0644)
+			if err != nil {
+				return kube.ClustersConfig{}, deferFn, err
+			}
+		}
+		configs.Clusters[c.Name] = c
+	}
+	return configs, deferFn, nil
+}
+
+// createClientGetter returns a function closure for creating the k8s client to interact with the cluster.
+// The returned function utilizes the user credential present in the request context.
+// The plugins just have to call this function passing the context in order to retrieve the configured k8s client
+func createClientGetter(serveOpts ServeOptions) (func(context.Context) (dynamic.Interface, error), error) {
+	// get the default rest incluster config for the kube.NewClusterConfig function
+	var inClusterConfig, err = rest.InClusterConfig()
+	if err != nil {
+		return nil, fmt.Errorf("unable to get inClusterConfig: %w", err)
+	}
+
+	// get the parsed kube.ClustersConfig from the serveOpts
+	config, err := getClustersConfigFromServeOpts(serveOpts)
+	if err != nil {
+		return nil, err
+	}
+
+	// return the closure fuction that takes the context, but preserving the required scope,
+	// 'inClusterConfig' and 'config'
+	return createClientGetterWithParams(inClusterConfig, serveOpts, config)
+}
+
+// createClientGetter takes the required params and returns the closure fuction.
+// it's splitted for testing this fn separately
+func createClientGetterWithParams(inClusterConfig *rest.Config, serveOpts ServeOptions, config kube.ClustersConfig) (func(context.Context) (dynamic.Interface, error), error) {
+
+	// return the closure fuction that takes the context, but preserving the required scope,
+	// 'inClusterConfig' and 'config'
+	return func(ctx context.Context) (dynamic.Interface, error) {
+		var err error
+		token, err := extractToken(ctx)
+		if err != nil {
+			return nil, status.Errorf(codes.Unauthenticated, "invalid authorization metadata: %v", err)
+		}
+
+		var client dynamic.Interface
+		if !serveOpts.UnsafeUseDemoSA {
+			restConfig, err := kube.NewClusterConfig(inClusterConfig, token, "default", config)
```

Comment:
Guessing you didn't mean to hard-code the "default" cluster here?

Reply:
Oops, leftover I didn't notice! It should retrieve the cluster from `ClustersConfig.KubeappsClusterName` (since we are not using cluster-scoped endpoints yet, afaik). Changing it straight away. Thanks!